### PR TITLE
[opentitantool] Enable Command Sequencing

### PR DESCRIPTION
**Pull request by cfrantz**
1. Add a `--exec` option that will allow multiple commands to run in
   sequence.

   Example:
     - Start the console with a timeout of zero (ie: force UART init).
     - Bootstrap `foo-code.bin` into the target.
     - Start the console in interactive mode to view the output.

   ```
   $ opentitantool --interface=cw310 \
       --exec="console -q -t 0" \
       --exec="bootstrap foo-code.bin" \
       console
   ```

Signed-off-by: Chris Frantz <cfrantz@google.com>

### CODE REVIEW

File: sw/host/opentitantool/src/main.rs [hunk 1]
```diff
@@ -112,6 +119,14 @@ fn main() -> Result<()> {
 
     let transport = backend::create(&opts.backend_opts)?;
 
+    if let Some(arg) = &opts.bootstrap {
+        let mut args = vec![OsString::from("bootstrap"); 1];
```
Comment (mcy) [inline sw/host/opentitantool/src/main.rs:123]: You can drop the `; 1`.

File: sw/host/opentitantool/src/main.rs [hunk 2]
```diff
@@ -112,6 +119,14 @@ fn main() -> Result<()> {
...
+        args.extend(shellwords::split(arg)?.iter().map(OsString::from));
+        let bootstrap = command::bootstrap::BootstrapCommand::from_iter(&args);
```
Comment (mcy) [inline sw/host/opentitantool/src/main.rs:125]: If you wanted to be really cool one way you could achieve this is
```rust
BootstrapCommand::from_iter(iter::once("bootstrap".to_string()).chain(shellwords::split(arg)?).map(OsString::from))
```
but that might be too wordy.

**Review by mcy — APPROVED** [review]

**Review by jesultra — APPROVED** [review]
I do not see anything wrong with adding this feature.

The only thing I could wish for was a more general way of doing a sequence of things. E.g. choose a particular strapping, adding a few explicitly set external GPIO signals, then bootstrapping a given image, while opening the console uart.  I guess that in general, you can just do a sequence of invocations of the tool from a script, but that could be said about bootstrapping before one other thing as well.

While thinking about this feature, if I use --bootstrap on a "console" command, is my UART console guaranteed to capture characters produced immediately upon the image booting?  Such a guarantee would be necessary for reliably test scripting.  Upholding the guarantee probably requires the serial port to be opened _before_ the reset pin is released.

**Comment by cfrantz** [comment]
> The only thing I could wish for was a more general way of doing a sequence of things. E.g. choose a particular strapping, adding a few explicitly set external GPIO signals, then bootstrapping a given image, while opening the console uart. I guess that in general, you can just do a sequence of invocations of the tool from a script, but that could be said about bootstrapping before one other thing as well.
> 
> While thinking about this feature, if I use --bootstrap on a "console" command, is my UART console guaranteed to capture characters produced immediately upon the image booting? Such a guarantee would be necessary for reliably test scripting. Upholding the guarantee probably requires the serial port to be opened _before_ the reset pin is released.

StructOpt doesn't "natively" support this sort of command sequencing, but I think we can support it.  I've added a commit to this PR which adds a `--sequence` flag.  When set, the program treats each positional arg as a command to be parsed (whereas the default behavior is to treat all of the positional args as a single command)

An example:
     - Start the console with a timeout of zero (ie: force UART init).
     - Bootstrap `foo-code.bin` into the target.
     - Start the console in interactive mode to view the output.

   ```
   $ opentitantool --interface=cw310 --sequence \
       "console -q -t 0" \
       "bootstrap foo-code.bin" \
       "console"
   ```

My concern with this approach is that structopt no longer knows about all of the subcommands at the root `Opt` structure.  If you issue the `help` command, it will list them out, but prior to this change, I believe it would also list hem out with the `--help` option.  This is, perhaps, not to big of a problem.

WDYT?

**Comment by jesultra** [comment]
I think the --sequence feature would be quite useful.  If we have to use "help" instead of "--help", that is perfectly fine with me.  (Particularly if we can make sure that --help displays a message to that effect.)

**Comment by cfrantz** [comment]
@jesultra After our discussion last week, I went with adding an `--exec` multi-arg to specific commands to parse and execute.

File: sw/host/opentitantool/src/main.rs [hunk 3]
```diff
@@ -45,6 +53,9 @@ struct Opts {
     #[structopt(long, default_value = "off")]
     logging: LevelFilter,
 
+    #[structopt(long, help = "Parse and execute the argument as a command")]
+    exec: Vec<String>,
```
Comment (jesultra) [inline sw/host/opentitantool/src/main.rs:57]: How is it that you tell structopt to expect `--exec "a" --exec "b"`, rather than `--exec "a" "b"` ?  I seem to remember the default being the latter, when dealing with the `--conf` option, it causing `--conf a.json bootstrap` to be parsed as two file names, rather than a file name and a command verb.

**Review by jesultra — APPROVED** [review]

File: sw/host/opentitantool/src/main.rs [hunk 4]
```diff
@@ -107,11 +118,32 @@ fn parse_command_line(opts: Opts, mut args: ArgsOs) -> Result<Opts> {
     Ok(opts)
 }
 
+fn execute<I>(args: I, opts: &Opts, transport: &TransportWrapper) -> Result<()>
```
Comment (mcy) [inline sw/host/opentitantool/src/main.rs:121]: Is there a particular reason this is factored out into a function? Is it to be able to catch `?` in the right way or something?
Probably worth documenting if so.

Comment (cfrantz) [reply sw/host/opentitantool/src/main.rs:121]: Originally, I had factored this out into a function because I was approaching command sequencing by collecting all positional args and making a decision about whether they need to be split with `shellwords` or could be passed as a set to `RootCommandHierarch::from_iter`.

As it is now, I think it makes the `for command in &opts.exec` loop look less gnarly.

I'm considering adding a readline-based interactive mode in the future, in which case, this function will also be useful for dispatching those commands.

File: sw/host/opentitantool/src/main.rs (hunk 3 above)
Comment (cfrantz) [reply sw/host/opentitantool/src/main.rs:57]: This is not well documented by structopt.  You need to use the `--exec="stuff"` form to get the desired behavior.  You are correct that `--exec "a" bootstrap` will get parsed into `exec=["a", "bootstrap"]`, whereas `--exec="a" bootstrap` will get parsed into `exec=["a"], command=BootStrapCommand()`.

I'll take a closer look at structopt and perhaps file an issue with them.

File: sw/host/opentitantool/src/main.rs (hunk 4 above)
Comment (cfrantz) [reply sw/host/opentitantool/src/main.rs:123]: Added a comment explaining that `execute` is a convenience function.

File: sw/host/opentitantool/src/main.rs (hunk 3 above)
Comment (cfrantz) [reply sw/host/opentitantool/src/main.rs:57]: After a bit more study of the structopt and clap documentation, I think we should add `number_of_values(1)` to any multi-args.  This forces the flag to be satisfied after consuming one argument without requiring the `--foo=bar` form, but allows flags which take a Vec<T> to be specified multiple times to supply additional values into the Vec.

See https://docs.rs/clap/2.34.0/clap/struct.Arg.html#method.number_of_values
